testing_app/test_driver/perf_test.dart

```diff
@@ -0,0 +1,123 @@
+// Copyright 2020 The Flutter team. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter_driver/flutter_driver.dart';
+import 'package:test/test.dart';
+
+void main() {
+  group('Testing App Performance Tests', () {
+    FlutterDriver driver;
+
+    // Connect to the Flutter driver before running any tests.
+    setUpAll(() async {
+      driver = await FlutterDriver.connect();
+    });
+
+    // Close the connection to the driver after the tests have completed.
+    tearDownAll(() async {
+      if (driver != null) {
+        await driver.close();
+      }
+    });
+
+    test('Scrolling test', () async {
+      // Create Finders that are used multiple times.
+      final listFinder = find.byType('ListView');
+      final firstItem = find.byValueKey('text_0');
+      final lastItem = find.byValueKey('text_49');
+
+      // Record a performance profile as the app scrolls through
+      // the list of items.
+      final scrollingTimeline = await driver.traceAction(() async {
+        // Scroll until the item to be found appears.
+        // Use dxScroll to scroll horizontally and dyScroll
+        // to scroll vertically.
+        await driver.scrollUntilVisible(
+          listFinder,
+          lastItem,
+          dyScroll: -1000.0,
+        );
+
+        // Check if the item contains the correct text.
+        expect(await driver.getText(lastItem), 'Item 49');
+
+        // Scroll back up all the way.
+        await driver.scrollUntilVisible(
+          listFinder,
+          firstItem,
+          dyScroll: 1000.0,
+        );
+
+        // Check if the item contains the correct text.
+        expect(await driver.getText(firstItem), 'Item 0');
+      });
+
+      // Convert the Timeline into a TimelineSummary that's easier to
```

Comment:
In contrast to what I said above about removing some of the inline comments, these are *really* good and definitely add to the codebase. This is the most complicated thing you're trying to teach, and the comments will make it easier for many readers to parse and understand this code.

Reply:
I think it might be even better to mention where the timeline is write to (i.e. `build/scrolling.timeline.json` and `build/scrolling.timeline_summary.json`.

Reply:
@CareF I think that's something I forgot to mention. Thanks for reminding me!